### ISSUE TITLE
Bug: @typescript-eslint/parser - missing types 

### ISSUE THREAD

**Issue by ivodolenc** [typescript-eslint/typescript-eslint#7279]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

parser

### Playground Link

-

### Repro Code

```typescript
-
```

### ESLint Config

```javascript
-
```

### tsconfig

```jsonc
-
```

### Expected Result

No errors.

### Actual Result

Cannot find module '@typescript-eslint/parser' or its corresponding type declarations.

<img width="944" src="https://github.com/typescript-eslint/typescript-eslint/assets/51852491/75ca309d-448f-4e9e-9d8b-da9a695685f3">

### Additional Info

Hi, I believe you need to specify `types` in `package.json` because when using the parser in `.ts` files it reports missing type declarations.

Simple fix (I tested this manually and it works as expected):

**@typescript-eslint/parser:  package.json**

```diff
{
+ "types": "./dist/index.d.ts",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "default": "./dist/index.js"
    },
    "./package.json": "./package.json"
  }
}
```

### Versions

"@typescript-eslint/parser": "^6.1.0"

**Comment by bradzacher** [comment]
You need to use `moduleResolution: node16`
https://www.typescriptlang.org/tsconfig#moduleResolution

Using `moduleResolution: node` ignores package.json exports and might allow you to write code that will crash at runtime.

**Comment by ivodolenc** [comment]
Hi @bradzacher, tnx for quick reply.

Yes, I am aware of that but `node16` is a story for itself which also complicates things.

Could you please consider adding that `types` line to `package.json` since it won't harm the plugin at all and won't break anything on your side?

I'm asking this because I am creating a custom configuration (new eslint flat config system) that I will publish as an `npm` package and later import into other projects.

As I mentioned earlier, when importing the `parser` into the `.ts` files, it reports an error that there are no types, which can be quite irritating.

---

> Using moduleResolution: node ignores package.json exports and might allow you to write code that will crash at runtime.

I also believe that nothing will crash at runtime...

Most large projects (to name a few: next, nuxt, rollup, vite...) use `node` moduleResolution and everything works fine.

---

Also, I came across a similar [issue](https://github.com/typescript-eslint/typescript-eslint/issues/7213#issuecomment-1641826507) and suppose the workaround will be to use `"moduleResolution": "bundler"` but it's not possible in every project.

**Comment by ivodolenc** [comment]
I also forgot to mention a quick fix from the user side.

Simply add a `/dist` suffix to the import path:

```ts
// No types error
import tsParser from '@typescript-eslint/parser/dist'
```

But I guess it would probably be best to handle this on the author/package side.

**Comment by bradzacher** [comment]
You can try that, but it will fail on any version of node we support:

```
> require('@typescript-eslint/parser/dist')
Uncaught:
Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: Package subpath './dist' is not defined by "exports" in E:\github\test\node_modules\@typescript-eslint\parser\package.json
    at __node_internal_captureLargerStackTrace (node:internal/errors:490:5)
    at new NodeError (node:internal/errors:399:5)
    at exportsNotFound (node:internal/modules/esm/resolve:361:10)
    at packageExportsResolve (node:internal/modules/esm/resolve:697:9)
    at resolveExports (node:internal/modules/cjs/loader:565:36)
    at Module._findPath (node:internal/modules/cjs/loader:634:31)
    at Module._resolveFilename (node:internal/modules/cjs/loader:1061:27)
    at Module._load (node:internal/modules/cjs/loader:920:27)
    at Module.require (node:internal/modules/cjs/loader:1141:19)
    at require (node:internal/modules/cjs/helpers:110:18) {
  code: 'ERR_PACKAGE_PATH_NOT_EXPORTED'
}
```

---

> Could you please consider adding that types line to package.json since it won't harm the plugin at all and won't break anything on your side?

We support minimum node16 and our packages define `package.json` exports to purposefully and intentionally control and limit our surface area.
We do not support node10 resolution because it emulates a runtime resolution that no longer works and will actively crash at runtime.

If you can't turn `node16` on in the root of your app, then the solution is to create a tsconfig file close to the file(s) that need our tooling so you can isolate the impact of the correct module resolution config.

If your project is that restricted that you can't do that - then consider using `patch-package` to change it locally - but note that you'd be working in an unsupported setup and thus in a "no warranty provided" environment.

**Comment by ivodolenc** [comment]
This is my `tsconfig.json`:

```json
{
  "compilerOptions": {
    "target": "ESNext",
    "module": "ESNext",
    "moduleResolution": "Node",
    "skipLibCheck": true,
    "strict": true,
    "allowJs": true,
    "noEmit": true,
    "resolveJsonModule": true,
    "allowSyntheticDefaultImports": true,
    "forceConsistentCasingInFileNames": true
  },
  "include": ["src"],
  "exclude": ["node_modules", "dist"]
}
```

This is my `package.json`:

```json
{
  "type": "module",
  "types": "./dist/index.d.ts",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.mjs"
    }
  }
}
```

Also, I'm using `rollup` to bundle my npm packages and nothing is broken.

```ts
// This works fine without ts errors
import tsParser from '@typescript-eslint/parser/dist'
```

The types property in package.json simply tells where to look for types.

You can have both types and exports, just like main, modules and exports, and nothing will be broken on your end, so no need to discuss further.

If you don't want to add this, fine, I understand, but to clarify, nothing will be broken.

As I mentioned, a lot of big projects work that way so it doesn't make sense to say it doesn't work or crashes at runtime.

Thanks for your time to look at this. Consider it resolved.
